### PR TITLE
5) rust: use AsRef/AsMut in compute interface

### PR DESCRIPTION
**Pull request by crop2000**
the current compute function signature looks like this:
```
	pub fn compute(&mut self, count: usize, inputs: &[&[FaustFloat]], outputs: &mut [&mut [FaustFloat]]){
```
this prevents us from passing a mutable reference as input buffer slice.
in the case one wants to chain dsps where the outputs of one dsp are used as the input of another it would be good if one can pass pass same slice that is used as the output of one dsp as input of another. but in that case one would pass a mutable reference. which is not possible with the current function signature.
this pr changes the functions to be more general. so that one can pass a mutable reference to the the input.
The new now looks like this:
```
	pub fn compute<I, O>(
		&mut self,
		count: usize,
		inputs: &[I],
		outputs: &mut [O],
	) where
		I: AsRef<[FaustFloat]>,
		O: AsMut<[FaustFloat]>,
	{
```
this change should be fully backwards compatible. (as first attemps have shown that needs some review and improvements of testing)

Because I found out that my earlier introduced compute_array interface is not contributing to the performance i removed it here again because it would make this change more complicated.

### CODE REVIEW

**Comment by crop2000** [comment]
@sletz this one is also ready

File: tests/impulse-tests/Makefile [hunk 1]
```diff
@@ -122,10 +122,7 @@ travis:
 #########################################################################
 # automatic github action test
 github_action:
-	$(MAKE) -f Make.rust outdir=rust/osec CARGOOPTIONS="" FAUSTOPTIONS="-I ../../libraries/ -double -os -ec -rnt -a archs/rust/architecture_osecrnt.rs"
```
Comment (sletz) [inline tests/impulse-tests/Makefile:125]: Why do you remove the tests here ? (I think it make sense to keep all of them)

Comment (crop2000) [reply tests/impulse-tests/Makefile:125]: The rational was to only keep tests that relate to a change in ci. I don't have a strong opinion on this.

**Comment by sletz** [comment]
You wrote **this change is fully backwards compatible** but `faust2portaudiorust` and `faust2jackrust` are broken after this change.  So what are existing projects supposed to do ?

**Comment by crop2000** [comment]
I made a mistake before because i thought i already implemented the most general interface.
Now i wrote a test to compare different interfaces and updated the pr accordingly.
[here](https://gist.github.com/crop2000/12af81bd2093fb43cacf6f90ef7af90d) is the test for documentation.
I also updated some architecture files so that the cover differently typed buffers to be passed to the compute function.

**Comment by crop2000** [comment]
I think i will update the PR one more time to match the function signature as i updated it in the pr description above.

**Comment by bluenote10** [comment]
>  @bluenote10 could I ask your "more expert than me" overview on this PR ?

Sure I can look into this in a bit more detail, but will probably only get to it next weekend.

**Comment by crop2000** [comment]
i just realized the function signature can be simplified with the same meaning:
```
pub fn compute(&mut self, count: usize, inputs: &[impl AsRef<[FaustFloat]>], outputs: &mut [impl  AsMut<[FaustFloat]>])
```
lets see if i can make this change before the review ... but functionally this should be equivalent to the current version.

**Comment by bluenote10** [comment]
I had a closer look now into this, and as far as I can see this should be all fine!

I was re-rerunning [all my benchmarks](https://github.com/bluenote10/RustFaustExperiments/tree/master/Benchmarks) before and after that change to get a better feeling of how the generated code looks, and if it has an impact on performance. For what I've seen it all looks fine to me, and I would assume that it likely produces the same assembly anyway.

**Comment by crop2000** [comment]
@sletz @bluenote10 thanks for having a look. I am also preparing benchmarks,  but havnt cleaned them up.

**Comment by sletz** [comment]
> @sletz @bluenote10 thanks for having a look. I am also preparing benchmarks, but havnt cleaned them up.

Should I wait for that before merging ?

**Comment by crop2000** [comment]
@sletz no need to wait for publishable benchmarks.

**Comment by sletz** [comment]
Merged, thanks.

**Pull request closed without merging**